File: linked_list/1-linked_list_print.js
// Write code to print a linked list forward and backward.

class LinkedList {

    constructor () {
        this.length = 0
        this.head = null
    }

    dumpHead () {
        console.log(`Head = ${this.head.cargo}`)
    }

    add (cargo) {
        let next = this.head
        this.head = new Node(cargo, next)
        this.length += 1
    }

    print () {
        console.log(`------------------print`)
        if (this.head) {
            this.head.print()
        }
    }

    printBack () {
        console.log(`------------------printBack`)
        if (this.head) {
            this.head.printBack()
        }
    }

}

class Node {

    constructor (cargo, next = null) {
        this.cargo = cargo
        this.next = next
    }

    print () {
        console.log(this.cargo)
        if (this.next) {
            this.next.print()
        }
    }

    printBack () {
        if (this.next) {
            this.next.printBack()
        }
        console.log(this.cargo)
    }

}

let list = new LinkedList()

list.add('Arivederci')
list.add('Flutchen krugen')
list.add('Guten morge')
list.add('Konnichonwa')
list.add('Hello')

list.dumpHead()

list.print()

list.printBack()

File: linked_list/1-linked_list_print.py
"""
Write code to print a linked list forward and backward.
"""

class LinkedList:

    def __init__(self):
        self.length = 0
        self.head = None

    def add(self, cargo):
        _next = self.head
        self.head = Node(cargo, _next)
        self.length += 1

    def print_forward(self):
        print('[', end='')
        if self.head:
            self.head.print_forward()
        print(']')

    def print_backward(self):
        print('[', end='')
        if self.head:
            self.head.print_backward()
        print(']')


class Node:

    def __init__(self, cargo=None, _next=None):
        self.cargo = cargo
        self.next = _next

    def print_forward(self):
        print(self, end=',' if self.next else '')
        if self.next:
            self.next.print_forward()

    def print_backward(self):
        if self.next:
            self.next.print_backward()
        print(self.cargo, end=',')

    def __str__(self):
        return str(self.cargo)


linked_list = LinkedList()

linked_list.add(1)
linked_list.add(2)
linked_list.add(3)

print('-' * 40)
linked_list.print_forward()

print('-' * 40)
linked_list.print_backward()
